### PR TITLE
fix: refine stability precedence to support standard prerelease sorting

### PR DESCRIPTION
**Pull request by sameerforge**
**Refinement of Metadata Precedence Logic**
This follow-up PR adjusts the build metadata comparison logic to ensure stability-aware precedence (GA > RC) only applies when comparing numeric segments to alphanumeric ones.

**Verified Improvements:**
**Natural Sorting:** vmware.10-fips now correctly ranks higher than vmware.9-fips.

**Stability:** 3.4.0+v1.33 (Stable) now correctly ranks higher than 3.4.0+v1.33-rc.2.

**Compatibility:** Standard pre-release ordering (e.g., 0.0.1-pre.1 < 0.0.1-rc.0) is preserved, fixing a regression found in vendir.

**Validation :** 
All 38+ existing unit tests, alongside new edge-case coverage Passed ✅
```
Current changeset Validation : 
➜  v4 git:(topic/sameerforge/follow-up-semver-fix) go test -v .                                                        
=== RUN   TestJSONMarshal
--- PASS: TestJSONMarshal (0.00s)
=== RUN   TestJSONUnmarshal
--- PASS: TestJSONUnmarshal (0.00s)
=== RUN   TestParseComparator
--- PASS: TestParseComparator (0.00s)
=== RUN   TestSplitAndTrim
--- PASS: TestSplitAndTrim (0.00s)
=== RUN   TestSplitComparatorVersion
--- PASS: TestSplitComparatorVersion (0.00s)
=== RUN   TestBuildVersionRange
--- PASS: TestBuildVersionRange (0.00s)
=== RUN   TestSplitORParts
--- PASS: TestSplitORParts (0.00s)
=== RUN   TestGetWildcardType
--- PASS: TestGetWildcardType (0.00s)
=== RUN   TestCreateVersionFromWildcard
--- PASS: TestCreateVersionFromWildcard (0.00s)
=== RUN   TestIncrementMajorVersion
--- PASS: TestIncrementMajorVersion (0.00s)
=== RUN   TestIncrementMinorVersion
--- PASS: TestIncrementMinorVersion (0.00s)
=== RUN   TestExpandWildcardVersion
--- PASS: TestExpandWildcardVersion (0.00s)
=== RUN   TestVersionRangeToRange
--- PASS: TestVersionRangeToRange (0.00s)
=== RUN   TestRangeAND
--- PASS: TestRangeAND (0.00s)
=== RUN   TestRangeOR
--- PASS: TestRangeOR (0.00s)
=== RUN   TestParseRange
--- PASS: TestParseRange (0.00s)
=== RUN   TestMustParseRange
--- PASS: TestMustParseRange (0.00s)
=== RUN   TestMustParseRange_panic
--- PASS: TestMustParseRange_panic (0.00s)
=== RUN   TestStringer
--- PASS: TestStringer (0.00s)
=== RUN   TestParse
--- PASS: TestParse (0.00s)
=== RUN   TestParseTolerant
--- PASS: TestParseTolerant (0.00s)
=== RUN   TestMustParse
--- PASS: TestMustParse (0.00s)
=== RUN   TestMustParse_panic
--- PASS: TestMustParse_panic (0.00s)
=== RUN   TestValidate
--- PASS: TestValidate (0.00s)
=== RUN   TestFinalizeVersionMethod
--- PASS: TestFinalizeVersionMethod (0.00s)
=== RUN   TestCompare
--- PASS: TestCompare (0.00s)
=== RUN   TestWrongFormat
--- PASS: TestWrongFormat (0.00s)
=== RUN   TestWrongTolerantFormat
--- PASS: TestWrongTolerantFormat (0.00s)
=== RUN   TestCompareHelper
--- PASS: TestCompareHelper (0.00s)
=== RUN   TestIncrements
--- PASS: TestIncrements (0.00s)
=== RUN   TestPreReleaseVersions
--- PASS: TestPreReleaseVersions (0.00s)
=== RUN   TestBuildMetaDataVersions
--- PASS: TestBuildMetaDataVersions (0.00s)
=== RUN   TestNewHelper
--- PASS: TestNewHelper (0.00s)
=== RUN   TestMakeHelper
--- PASS: TestMakeHelper (0.00s)
=== RUN   TestFinalizeVersion
--- PASS: TestFinalizeVersion (0.00s)
=== RUN   TestEdgeCases_NaturalSortAndMetadata
--- PASS: TestEdgeCases_NaturalSortAndMetadata (0.00s)
=== RUN   TestSort
--- PASS: TestSort (0.00s)
=== RUN   TestScanString
--- PASS: TestScanString (0.00s)
PASS
ok      github.com/carvel-dev/semver/v4 0.502s
```

**Validation with Vendir Test :** 
Tests Passed ✅
```
➜  vendir git:(develop) ✗ go test -v ./pkg/vendir/versions/                                   

=== RUN   TestSemverOrder
--- PASS: TestSemverOrder (0.00s)
=== RUN   TestSemverFilter
--- PASS: TestSemverFilter (0.00s)
=== RUN   TestSemverWithoutPrereleases
--- PASS: TestSemverWithoutPrereleases (0.00s)
=== RUN   TestSemverWithPrereleases
--- PASS: TestSemverWithPrereleases (0.00s)
=== RUN   TestSemverWithPrereleaseIdentifiers
--- PASS: TestSemverWithPrereleaseIdentifiers (0.00s)
=== RUN   TestSemverWithBuildMetadata
--- PASS: TestSemverWithBuildMetadata (0.00s)
=== RUN   TestSemverWithBuildMetadataAndConstraint
--- PASS: TestSemverWithBuildMetadataAndConstraint (0.00s)
=== RUN   TestHighestVersionWithConstraints
--- PASS: TestHighestVersionWithConstraints (0.00s)
PASS
ok      carvel.dev/vendir/pkg/vendir/versions   (cached)
```

### CODE REVIEW

**Comment by sameerforge** [comment]
@joaopapereira Please review this follow-up change for https://github.com/carvel-dev/semver/pull/5

**Comment by sameerforge** [comment]
@joaopapereira Please review this follow-up change for https://github.com/carvel-dev/semver/pull/5

File: v4/semver.go [hunk 1]
```diff
@@ -472,19 +472,21 @@ func (v versionExtension) Compare(o versionExtension) int {
 	// Identify segments containing pre-release identifiers (rc, alpha, beta).
 	isPreReleaseLabel := func(s string) bool {
 		sl := strings.ToLower(s)
-		return strings.Contains(sl, "rc") || (strings.Contains(sl, "-") &&
-			(strings.Contains(sl, "alpha") || strings.Contains(sl, "beta")))
+		if strings.Contains(sl, "rc") {
+			return true
+		}
+		return strings.Contains(sl, "-") && (strings.Contains(sl, "alpha") || strings.Contains(sl, "beta"))
```
Comment (ramineni) [inline v4/semver.go]: 1. we are only checking for '-' for alpha only, not for beta or rc . We have to check for '-' as pre-requisite for all the pre release labels.

2. why are we restricting to check only rc , alpha, beta ? there can be other labels (pre etc) which can exist?  we can do string comparison to evaluate them?

File: v4/semver_test.go [hunk 2]
```diff
@@ -597,6 +597,15 @@ func TestEdgeCases_NaturalSortAndMetadata(t *testing.T) {
 
 		// 3. Mixed Alphanumeric: abc2 should be less than abc10
 		{"1.0.0+abc2", "1.0.0+abc10", -1, "Natural sort: abc2 should be less than abc10"},
+
+		// 4. Vendir Compatibility (The new case pre < rc)
+		{"0.0.1-pre.1", "0.0.1-rc.0", -1, "Vendir: pre < rc alphabetically"},
+
+		// 5. Standard SemVer: Numeric segments are always less than alphanumeric ones
+		{"1.0.0-alpha.1", "1.0.0-alpha.beta", -1, "Standard: numeric < alphanumeric"},
```
Comment (ramineni) [inline v4/semver_test.go]: this is the comparison for alphanumeric vs alpha numeric .

File: v4/semver_test.go [hunk 3]
```diff
@@ -597,6 +597,15 @@ func TestEdgeCases_NaturalSortAndMetadata(t *testing.T) {
...
+
+		// 6. Stability Check: Ensure alpha/beta without hyphens follow standard rules
+		{"1.0.0-beta", "1.0.0-rc", -1, "Standard: beta < rc alphabetically"},
```
Comment (ramineni) [inline v4/semver_test.go]: this example shows with '-' ?

File: v4/semver.go (hunk 1 above)
Comment (sameerforge) [reply v4/semver.go]: Updated. Logic now generically uses hyphens to detect stability label.

File: v4/semver_test.go (hunk 2 above)
Comment (sameerforge) [reply v4/semver_test.go]: Corrected. Updated the description to clarify this compares a numeric segment (1) vs an alphanumeric segment (beta) per SemVer precedence.

File: v4/semver_test.go (hunk 3 above)
Comment (sameerforge) [reply v4/semver_test.go]: Updated Description. The separator hyphen is not part of the segment itself

**Comment by sameerforge** [comment]
@joaopapereira Please review this follow-up change for https://github.com/carvel-dev/semver/pull/5

File: v4/semver.go [hunk 4]
```diff
@@ -470,21 +470,30 @@ func (v versionExtension) Compare(o versionExtension) int {
...
-			(strings.Contains(sl, "alpha") || strings.Contains(sl, "beta")))
+	isPreReleaseLabel := func(sl string) bool {
+		s := strings.ToLower(sl)
+		if !strings.Contains(s, "-") {
+			return false
+		}
+		// Only flag as pre-release if it contains these specific markers
+		preReleaseMarkers := []string{"alpha", "beta", "rc", "pre", "dev"}
+		for _, marker := range preReleaseMarkers {
+			if strings.Contains(s, marker) {
```
[the remark this answers is not in the document]
Comment (sameerforge) [reply v4/semver.go]: Made the suggested change.

**Review by ramineni — APPROVED** [review]

**Comment by sameerforge** [comment]
@joaopapereira PR is ready for your review now.

**Comment by sameerforge** [comment]
> 
> Comments suppressed due to low confidence (1)
> **v4/semver.go:510**
> 
> * The logic for comparing alphanumeric segments no longer includes stability-aware precedence for pre-release markers. This breaks test case "3.4.0+v1.33-rc.2" vs "3.4.0+v1.33", where the build metadata segment "v1.33-rc" should be less than "v1.33" (GA should rank higher than RC). However, with the current code, naturalLess("v1.33-rc", "v1.33") returns false because "v1.33-rc" has more chunks than "v1.33" after the common prefix ["v", "1", ".", "33"], causing the comparison to return 1 instead of -1. The removed string-vs-string pre-release marker comparison needs to be reintroduced for alphanumeric segments containing pre-release identifiers, or the test expectation needs to be updated to reflect that RC is now considered greater than GA in build metadata.
> 

All the Tests are passing as expected, along with the mentioned test case "3.4.0+v1.33-rc.2" vs "3.4.0+v1.33". The stability logic correctly prioritizes GA over RC before the natural sort fallback; this appears to be an AI false positive.

**Comment by sameerforge** [comment]
@joaopapereira Please review the final changes.

**Comment by sameerforge** [comment]
@joaopapereira  Please review the changes.

**Comment by sameerforge** [comment]
@joaopapereira Please review the changes.

**Review by joaopapereira — APPROVED** [review]
LGTM
